### PR TITLE
JAMES-1856 Upgrade to bouncy castle 1.52 jdk15on

### PR DESCRIPTION
**Pull request by chibenwa**
I also soved some IntelliJ warnings inside crypto mailets

### CODE REVIEW

**Comment by chibenwa** [comment]
test this please

**Comment by guillaumebadin** [comment]
I Have a bug with crypto mailet : 

java.lang.NoSuchMethodError: org.bouncycastle.cms.RecipientId.match(Ljava/security/cert/Certificate;)Z
	at org.apache.james.transport.mailet.SMIMEDecrypt.service(SMIMEDecrypt.java:122)

The method RecipientId.match(obj) on line 122 SMIMEDecrypt launch a java.lang.NoSuchMethodError

Does we need à specific compilation jdk ? I am with jdk8 !

**Comment by chibenwa** [comment]
Hi @guillaumebadin ,

I started working on your issue. 

Does you still observe this behaviour on this branch, or was it on master ? It might be because crypto-mailets were not updated during global bouncyCastle update (not the same artifact). As different versions of jars are at the same time on the classpath, behaviour might be random. This is what this PR try to address : align versions of bouncy catsle.

From thursday, I will start working to introduce some integration testing for crypto mailets, and will notify you as it get merged.

Best regards.

**Comment by guillaumebadin** [comment]
Hello, 
That's ok I didn't change the global dependency on java server. 
`  
 <dependency>
   <groupId>org.apache.james</groupId>
   <artifactId>apache-mailet-crypto</artifactId>
   <version>3.0.0-beta6-SNAPSHOT</version>
</dependency>
`

**Comment by chibenwa** [comment]
You are extending crypto mailet ?

Nice to ear this.

However, you might need to **mvn install** this version of crypto mailets to see changes it introduced. Is that what you did ?

**Comment by guillaumebadin** [comment]
@chibenwa Yes I am extending crypto mailet.

For information, the match method line 122 in SMIMEDecrypt.java doesn't work for me.
I juste test the serial number of the certificate and it works !
  if (id.getSerialNumber().equals(keyHolder.getCertificate().getSerialNumber())) {

I think the buncycastle match is too hard or it doesn't work at all.

File: mailet/crypto/src/main/java/org/apache/james/transport/mailets/AbstractSign.java [hunk 1]
```diff
@@ -606,7 +615,18 @@ protected boolean isOkToSign(Mail mail) throws MessagingException {
         return !isAlreadySigned;
 
     }
-    
+
+    private String getUsername(MailAddress mailAddress) {
```
Comment (chibenwa) [inline mailet/crypto/src/main/java/org/apache/james/transport/mailets/AbstractSign.java]: Will use UsersRepository::getUser after rebase on https://github.com/linagora/james-project/pull/461

File: mailet/crypto/src/main/java/org/apache/james/transport/mailets/AbstractSign.java [hunk 2]
```diff
@@ -576,7 +576,8 @@ protected boolean isOkToSign(Mail mail) throws MessagingException {
         }
         
         // The sender is the postmaster?
-        if (getMailetContext().getPostmaster().equals(reversePath)) {
+        if (getMailetContext().getPostmaster() != null &&
```
Comment (mbaechler) [inline mailet/crypto/src/main/java/org/apache/james/transport/mailets/AbstractSign.java]: Objects.equal is shorter

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/crypto/SMIMESignIntegrationTest.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            calmlyAwait.atMost(Duration.ONE_MINUTE).until(messageSender::messageHasBeenSent);
+            calmlyAwait.atMost(Duration.ONE_MINUTE).until(() -> imapMessageReader.userReceivedMessage(recipient, PASSWORD));
+
+            assertThat(imapMessageReader.readFirstMessageInInbox(recipient, PASSWORD))
+                .containsSequence("Content-Description: S/MIME Cryptographic Signature");
+        }
+    }
+
+    @Test
+    public void NonAuthenticatedMessagesShouldBeSigned() throws Exception {
```
Comment (mbaechler) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/crypto/SMIMESignIntegrationTest.java]: shouldNot ?

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/SMTPMessageSender.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+        AuthenticatingSMTPClient smtpClient = new AuthenticatingSMTPClient();
+        smtpClient.connect(ip, port);
+        smtpClient.auth(AuthenticatingSMTPClient.AUTH_METHOD.PLAIN, username, password);
+        return new SMTPMessageSender(smtpClient, senderDomain);
+    }
+
+    private final SMTPClient smtpClient;
+    private final String senderDomain;
+
+    public SMTPMessageSender(SMTPClient smtpClient, String senderDomain) {
```
Comment (aduprat) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/utils/SMTPMessageSender.java]: private

File: mailet/crypto/src/main/java/org/apache/james/transport/mailets/AbstractSign.java [hunk 5]
```diff
@@ -564,37 +564,45 @@ protected boolean isOkToSign(Mail mail) throws MessagingException {
...
         // Is it a bounce?
         if (reversePath == null) {
+            log("Can not sign : no sender");
             return false;
         }
         
         String authUser = (String) mail.getAttribute("org.apache.james.SMTPAuthUser");
         // was the sender user SMTP authorized?
         if (authUser == null) {
+            log("Can not sign mail for sender " + mail.getSender() + " as he is not a SMTP authenticated user");
```
Comment (rouazana) [inline mailet/crypto/src/main/java/org/apache/james/transport/mailets/AbstractSign.java]: please add single quotes around mail.getSender()

File: mailet/crypto/src/main/java/org/apache/james/transport/mailets/AbstractSign.java [hunk 6]
```diff
@@ -564,37 +564,45 @@ protected boolean isOkToSign(Mail mail) throws MessagingException {
...
         if (getMailetContext().getPostmaster().equals(reversePath)) {
             // should not sign postmaster sent messages?
             if (!isPostmasterSigns()) {
+                log("Can not sign mails for postmaster");
                 return false;
             }
         } else {
             // is the reverse-path user different from the SMTP authorized user?
             if (!reversePath.getLocalPart().equals(authUser)) {
+                log("SMTP logged in as " + authUser + " but pretend to be sender " + mail.getSender());
```
Comment (rouazana) [inline mailet/crypto/src/main/java/org/apache/james/transport/mailets/AbstractSign.java]: idem with authUser

File: mailet/crypto/src/main/java/org/apache/james/transport/mailets/AbstractSign.java [hunk 7]
```diff
@@ -564,37 +564,45 @@ protected boolean isOkToSign(Mail mail) throws MessagingException {
         
         // Is it a bounce?
         if (reversePath == null) {
+            log("Can not sign : no sender");
```
Comment (rouazana) [inline mailet/crypto/src/main/java/org/apache/james/transport/mailets/AbstractSign.java]: no space before ':' in English

File: server/mailet/integration-testing/src/test/java/org/apache/james/mailets/crypto/SMIMESignIntegrationTest.java [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+    }
+
+    @Test
+    public void authenticatedMessagesShouldBeSigned() throws Exception {
+        jamesServer.getServerProbe().addDomain(DEFAULT_DOMAIN);
+        String from = "user@" + DEFAULT_DOMAIN;
+        String recipient = "user2@" + DEFAULT_DOMAIN;
+        jamesServer.getServerProbe().addUser(from, PASSWORD);
+        jamesServer.getServerProbe().addUser(recipient, PASSWORD);
+        jamesServer.getServerProbe().createMailbox(MailboxConstants.USER_NAMESPACE, recipient, "INBOX");
```
Comment (rouazana) [inline server/mailet/integration-testing/src/test/java/org/apache/james/mailets/crypto/SMIMESignIntegrationTest.java]: These lines could be in setup, no?

**Pull request merged**